pyanaconda/modules/subscription/subscription.py

```diff
@@ -174,3 +254,99 @@ def set_system_purpose_data(self, system_purpose_data):
+        if request_type not in SUBSCRIPTION_REQUEST_VALID_TYPES:
+            raise InvalidValueError(
+                "Invalid subscription request type set '{}'".format(request_type)
+            )
+
+    @property
+    def subscription_request(self):
+        """Subscription request.
+
+        A DBus structure holding data to be used to subscribe the system.
+
+        :return: subscription request DBus structure
+        :rtype: DBusData instance
+        """
+        # Return a deep copy of the subscription request that
+        # has also been cleared of private data.
+        # Thankfully the secret Dbus structures modules
+        # has the get_public_copy() method that does just
+        # that. It creates a deep copy & clears
+        # all SecretData and SecretDataList instances.
+        return get_public_copy(self._subscription_request)
+
+    def set_subscription_request(self, subscription_request):
+        """Set a subscription request.
+
+        Set the complete DBus structure containing subscription
+        request data.
+
+        :param subscription_request: subscription request structure to be set
+        :type subscription_request: DBus structure
+        """
+        self._replace_current_subscription_request(subscription_request)
+        self.subscription_request_changed.emit()
+        log.debug("A subscription request set: %s", str(self._subscription_request))
+
+    def _replace_current_subscription_request(self, new_request):
+        """Replace current subscription request without loosing sensitive data.
+
+        We need to do this to prevent blank SecretData & SecretDataList instances
+        from wiping out previously set secret data. The instances will be blank
+        every time a SubscriptionRequest that went through get_public_copy() comes
+        back with the secret data fields unchanged.
+
+        So what we do is depends on type of the incoming secret data:
+
+        - SECRET_TYPE_NONE - use structure from new request unchanged,
+                             clearing previously set data (if any)
+        - SECRET_TYPE_HIDDEN - secret data has been set previously and
+                               cleared when SubscriptionRequest was sent out;
+                               put secret data from current request to the
+                               new one to prevent it from being lost
+                               (this will also switch the secret data
+                                instance to SECRET_TYPE_TEXT so that
+                                the Subscription module can read it
+                                internally)
+        - SECRET_TYPE_TEXT - this is new secret entry, we can keep it as is
+        """
+        current_request = self._subscription_request
+        # Red Hat account password
+        if new_request.account_password.type == SECRET_TYPE_HIDDEN:
+            new_request.account_password.set_secret(
```

Comment:
In this case, I would set `new_request.account_password = copy.deepcopy(current_request.account_password)`, because the hidden secret has no useful data. You shouldn't use `set_secret`, because it will set the secret to an empty string if the type of the current secret is `NONE`.

Reply:
Good point, changed! :)